src/observer.js

```diff
@@ -138,26 +138,104 @@ const reactiveMixin = {
         let isForcingUpdate = false
 
         function makePropertyObservableReference(propName) {
-            let valueHolder = this[propName]
-            const atom = new Atom("reactive " + propName)
+            let currentValue = this[propName]
+            let currentValueKeys = null == currentValue ? [] : Object.keys(currentValue)
+            const changeKeysAtom = new Atom("this." + propName + " [change keys]")
+            const individualAtoms = {}
+            const individualValues = {}
+
+            /**
+             * In-place convert properties to getter+setter
+             */
+            function convertStorageToReactive(storage) {
+                if (null == storage) {
+                    return storage
+                }
+                // props object is frozen, so we need to clone it
+                // we can do it safely inside componentWillMount (react won't warn us about it)
+                // btw, state object can not be cloned here, (react warns about direct state modification), however
+                // state object is not frozen, so we do not need to clone it to instrument
+                if (Object.isFrozen && Object.isFrozen(storage)) {
```

Comment:
Hmm, if we implemented this using Proxies, it would all be lot simpler I guess? I think that could be fine as well, only introduce this feature in the next mobx-react that accompanies MobX 5

Reply:
Unfortunatelly, we can't wrap state to proxy here, because react will check this and warn (proxied object !== original object). But for frozen props, of course, it could be applied.